### PR TITLE
WIP: RAM usage reduction

### PR DESCRIPTION
**Pull request by promovicz**
WORK IN PROGRESS! DO NOT MERGE YET!

The OpenPGP applet currently uses two large buffers, "tmp" and "buffer". The total RAM usage of the applet makes it impossible to install and use it on most common JavaCard models. More specifically only a few JCOP cards have enough memory to run the applet.

After some research I found that "tmp" can be eliminated entirely. This patch series is a work in progress to eliminate it entirely, which in itself is working on my branch.

The "tmp" buffer is used in four places:
- As a serialization buffer in sendPublicKey. This usage can be eliminated by using a better serialization algorithm that predicts memory size before starting to serialize.
- As an operational buffer in computeDigitalSignature, decipher and internalAuthenticate. This usage can be eliminated by recognizing that "buffer" is always large enough since the actual message size is always 256 for RSA-based operations with 2048bit keys. The current buffer would also be large enough for 4096bit operations, should they become supported.

My plan is to add additional size checks for added safety.

### CODE REVIEW

**Comment by promovicz** [comment]
I would of course love your comments on this patch and an indication on whether you would merge it once completed.

File: applet/src/openpgpcard/OpenPGPApplet.java [hunk 1]
```diff
@@ -1273,49 +1278,64 @@ private void importKey(APDU apdu) {
...
+		// child tag 0x82 (Exponent)
+		buffer[offset++] = (byte)0x82;
+		offset += putLength(buffer, offset, expLen);
+		offset += pubkey.getExponent(buffer, offset);
 
+		// done
 		return offset;
 	}
 
+	private short sizeLength(short length) {
```
Comment (klali) [inline applet/src/openpgpcard/OpenPGPApplet.java]: To me it looks like this function is doing the same thing as getLengthBytes()

**Comment by klali** [comment]
Hello,

Without looking to deeply I'd say that yes, we'd welcome a change like this. Let us know when you believe it's ready to be reviewed and merged.

/klas

**Comment by jonathancross** [comment]
@promovicz Is this abandoned?

**Comment by dogtopus** [comment]
Apparently someone else gave it a shot (8f7353dc)

**Pull request closed without merging**